### ISSUE TITLE
Doesn't upload files larger than 5 Gig to Swift

### ISSUE THREAD

**Issue by MartinPaulo** [gvlproject/cloudbridge#35]
If I use cloudbridge to upload a file larger than 5 Gig in size to Swift it eventually fails.

I've walked through the code and this is what I see happening: 
The `cloudbridge` `OpenStackBucketObject` `upload_from_file` method is called. This in turn eventually lands up calling the OpenStack `swiftclient.client.py` `Connection` `put_object` method: which is a low level method that does not attempt to break files > 5Gig in size into smaller segments. Hence when you hit the 5 Gig Swift wall it ends in tears..

The OpenStack utility class `SwiftService` has an upload method that will chunk files if you set the `segment_size` option. Thus this method will handle files > 5 Gig with aplomb.  I've tested this with some rough and ready sample code. I think this class should be delegated to rather than the `Connection` class when data is being uploaded...

**Comment by MartinPaulo** [comment]
I've done some further investigation. `SwiftService` is a multithreaded class, and it uses a connection factory method to make sure that each thread can create connections on demand. Rather than trying to copy settings across to this factory method from the cloudbridge openstack provider, it would be better to override this factory method so that the `_connect_swift` factory was called instead to mint connections on demand.

This can be done by simply overriding the `swiftclient.service` module method that mints the connections in the cloudbridge `OpenStackBucketObject` as follows:

```python
 swiftclient.service.get_conn = self._provider._connect_swift
```

It would also be good if the `_connect_swift` method could accept and use the options that  `SwiftService` class passes to its connection factory.

**Comment by MartinPaulo** [comment]
The `SwiftService` constructor builds its options from a grab bag of global, local and passed in options: and that it the filters out the options not needed for a connection by the way in which it calls its connection factory method. It does something along the lines of the following:

```python
class FakeConnection:
    def __init__(self, var1=None, var2=None, var3=None):
        self.var1 = var1
        self.var2 = var2
        self.var3 = var3

def get_fake_conn(f_options):
    return FakeConnection(
        f_options['var1'],
        f_options['var2'],
    )

grab_bag_of_options = {
    'var1': '1',
    'A': 'A',
    'var2': '2',
    'B': 'B',
}

connection = get_fake_conn(grab_bag_of_options)
```

So if the `_connect_swift` module is to accept options from the `SwiftService` instance it will also need to filter out options not used by the `Connection` constructor.

**Comment by MartinPaulo** [comment]
Further, it turns out that the delete of an uploaded object as implemented will fail if the object is larger than 5 Gig. But what do you know: the `SwiftService` class transparently handles the delete of segmented objects as well. Hence the `delete` method on the `OpenStackBucketObject` should also delegate to an instance of the `SwiftService` class.

**Comment by MartinPaulo** [comment]
I have created and tested a fix for this, and uploaded it as a pull request.

Some notes:

* My fix works on hard 5Gig boundaries. The OpenStack documentation states that segment sizes, and also the size of the files being uploaded, might work better on some installations of Swift if they are smaller. Hence this size limit should possibly be set as an option.
* The upload of non file data is not affected. So it will still fail if the data size is larger than 5 Gig. However, the `SwiftService` upload method expects a stream: and the current upload method is not limited to streams. Hence making it work would require wrapping non stream data with an appropriate stream. And the caller would have to set the `segment_size` option in advance, which would imply the base class contract would be broken. So I opted to leave it as is.
* I made the `connect_swift` method on the provider public and added the ability to accept options that are passed in by the `SwiftService`. These options are cloned and stripped of key/values that are not in the Connection signature. Also, the `os_options` value is set to none, as if it is set keystone v2 here at NeCTAR objects most strongly :(
